EP-11173 : OpenTelemetry Tracing Support (#11371)
kgateway-dev/kgateway: design/11173.md

## Motivation

Users are using the envoy OpenTelemetry tracer in production (https://github.com/envoyproxy/envoy/issues/24672#issuecomment-2592932718) and this functionality should be exposed through kgateway.
Kgateway currently supports writing envoy access logs to a file or sending them to an upstream gRPC service, however envoy also supports sending access logs to an OTEL collector which should be exposed to the user.

## Goals

- Users should be able to configure an otel tracing config via kgateway
- Users should be able to configure an otel access log config via kgateway

## Non-Goals

- Deploying an otel collector as part of the kgateway helm chart

## Implementation Details

### Configuration

A user should be able to configure an OTel tracing and accessLog config for a Gateway via the `HTTPListenerPolicy`:
```
apiVersion: gateway.kgateway.dev/v1alpha1
kind: HTTPListenerPolicy
  name: tracing-config
spec:
  tracing:
    openTelemetryConfig:
      grpcService:
        serviceName: tracing-service
        backendRef:
          name: gateway-proxy-tracing-server
          namespace: default
          port: 8083

>     accessLog:

Author: oops, need to update the indent. Thanks for catching this!

}

// OpenTelemetryTracingConfig represents the top-level Envoy's OpenTelemetry tracer.
// See here for more information: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/trace/v3/opentelemetry.proto.html
type OpenTelemetryTracingConfig struct {
    // Send traces to gRPC service
    // +kubebuilder:validation:Required
    GrpcService *GrpcService `json:"grpcService,omitempty"`

    // The name for the service. This will be populated in the ResourceSpan Resource attributes
    // +kubebuilder:validation:Required
    ServiceName string `json:"serviceName,omitempty"`
}
```
The existing AccessLog struct will be modified to add support for OTel access logs
```go
type AccessLog struct {
    // Send access logs to an OTel collector
    OpenTelemetryConfig OpenTelemetryAccessLogConfig `json:"openTelemetryConfig,omitempty"`
}

type OpenTelemetryAccessLogConfig struct {
    // Send access logs to gRPC service
    // +kubebuilder:validation:Required
    GrpcService *GrpcService `json:"grpcService,omitempty"`

    // name of log stream
    // +kubebuilder:validation:Required
    LogName string `json:"logName"`

> }

Author: Yes, thats right. I'll update the docs

# EP-11173: OpenTelemetry Tracing & Access Log Support

* Issue: [#11173](https://github.com/kgateway-dev/kgateway/issues/11173)
* Issue: [#11226](https://github.com/kgateway-dev/kgateway/issues/11226)

## Background

Enable OpenTelemetry (OTel) access logging and tracing capabilities to obtain visibility and track requests as they pass through the API gateway.
This allows users to obtain visibility and track requests as they pass through the API gateway to distributed backends, such as services, databases, or other endpoints.

## Motivation

Users are using the envoy OpenTelemetry tracer in production (https://github.com/envoyproxy/envoy/issues/24672#issuecomment-2592932718) and this functionality should be exposed through kgateway.

> Kgateway currently supports writing envoy access logs to a file or sending them to an upstream gRPC service, however envoy also supports sending access logs to an OTEL collector which should be exposed to the user.

Reviewer: ```suggestion
Kgateway currently supports writing envoy access logs to a file or sending them to an upstream gRPC service, however Envoy also supports sending access logs to an OTel Collector which should be exposed to the user.
```

nit

## Goals

> - Users should be able to configure an otel tracing config via kgateway

Reviewer: ```suggestion
- Users should be able to configure an OTel tracing via kgateway
```

> - Users should be able to configure an otel access log config via kgateway

Reviewer: ```suggestion
- Users should be able to configure an OTel access logging via kgateway
```

## Non-Goals

> - Deploying an otel collector as part of the kgateway helm chart

Reviewer: ```suggestion
- Deploying an OTel Collector as part of the kgateway Helm chart
```

type ResourceDetector string
```
The existing AccessLog struct will be modified to add support for OTel access logs
```go
type AccessLog struct {
    // Send access logs to an OTel collector
    OpenTelemetryConfig OpenTelemetryAccessLogConfig `json:"openTelemetryConfig,omitempty"`
}

type OpenTelemetryAccessLogConfig struct {
    // Send access logs to gRPC service
    // +kubebuilder:validation:Required
    GrpcService *GrpcService `json:"grpcService,omitempty"`

    // name of log stream
    // +kubebuilder:validation:Required
    LogName string `json:"logName"`

    // OpenTelemetry LogResource fields, following Envoy access logging formatting.
    // +kubebuilder:validation:Optional
    Body string `json:"body,omitempty"`

    // If specified, Envoy will not generate built-in resource labels like log_name, zone_name, cluster_name, node_name.
  	// +kubebuilder:validation:Optional
    DisableBuiltinLabels bool `json:"disableBuiltinLabels,omitempty"`

    // Additional attributes that describe the specific event occurrence.
  	// +kubebuilder:validation:Optional
    Attributes opentelemetry.io.proto.otlp.common.v1.KeyValueList `json:"attributes,omitempty"`
}
```

### Plugin

> The existing `internal/kgateway/extensions2/plugins/httplistenerpolicy/httplistener_plugin.go` performs the access log translation

Reviewer: The phrasing is a bit weird?

can be updated to support OTel access log and tracing.

### Controllers

> No new controllers will be required for this feature. The existing plugin will be updated to support this .

Reviewer: ```suggestion
No new controllers will be required for this feature. The existing plugin will be updated to support this.
```

### Deployer

The CRDs will be updated with the new fields along with the required validation

### Test Plan

The testing plan will include unit and kubernetes end to end tests to verify the feature

## Alternatives

- The tracing config can be defined in the `GatewayExtension`, however this could lead to invalid configs where a user might inadvertently apply it where it is not supported, eg: an HTTPRoute.

## Open Questions

- Which detectors should be supported ?

> - Should samplers be supported ?

Reviewer: I'd include samplers to provide better out-of-box UX.

# EP-11173: OpenTelemetry Tracing & Access Log Support

* Issue: [#11173](https://github.com/kgateway-dev/kgateway/issues/11173)
* Issue: [#11226](https://github.com/kgateway-dev/kgateway/issues/11226)

## Background

Enable OpenTelemetry (OTel) access logging and tracing capabilities to obtain visibility and track requests as they pass through the API gateway.
This allows users to obtain visibility and track requests as they pass through the API gateway to distributed backends, such as services, databases, or other endpoints.

## Motivation

Users are using the envoy OpenTelemetry tracer in production (https://github.com/envoyproxy/envoy/issues/24672#issuecomment-2592932718) and this functionality should be exposed through kgateway.
Kgateway currently supports writing envoy access logs to a file or sending them to an upstream gRPC service, however Envoy also supports sending access logs to an OTel Collector which should be exposed to the user.

> 

Reviewer 2: Maybe also mention some addition context from https://github.com/kgateway-dev/kgateway/issues/11226
- Brief overview of current FilelogReceiver approach and the tradeoffs with this approach
- Specific benefits of exposing Envoy OTLP Access Log extension mentioned in the issue and what they mean (easier configuration, flexibility wrt OTEL collector deployments, control over logs)

## Goals

- Users should be able to configure OTel tracing via kgateway
- Users should be able to configure OTel access logging via kgateway

> 

Reviewer 2: - Provide documentation and example configurations for integrating kgateway with popular observability platforms (e.g., Jaeger, Datadog).

Author: Thanks, added

> 

Author: Thanks, added

> 

Reviewer 2: Is there a section for the OTEL Collectors need to run as a Daemonset vs. Deployment to support the case where the gateway-proxy might be scheduled on a different node than the OTEL Collector from https://github.com/kgateway-dev/kgateway/issues/11226

> 

Reviewer: OTel Collector as a Deployment can scrape workloads cross-node too. We can go into more details when focusing on metrics, but this EP is logging/tracing specific, and for these use-cases Deployment is the default form factor.

> 

Reviewer 2: Yep, it was referenced in https://github.com/kgateway-dev/kgateway/issues/11226 in the context of AccessLogging (in the context of how to guarantee access to the gateway-proxies logs, as the gateway-proxy might be scheduled on a different node than the OTEL Collector)

> 

Author: If envoy sends the logs / traces to a kuberentes service and not a pod itself, and if configured correctly, Kubernetes ensures that they reach the appropriate pod irrespective of the node on which it resides.

## Non-Goals

- Deploying an OTel Collector as part of the kgateway Helm chart

> 

Reviewer 2: - Providing opinionated configurations with specific observability vendors or backends out of the box (integrations with kgateway should be vendor agnostic).

Author: Thanks, added

## Implementation Details

### Configuration

A user should be able to configure an OTel tracing and accessLog config for a Gateway via the `HTTPListenerPolicy`:
```
apiVersion: gateway.kgateway.dev/v1alpha1
kind: HTTPListenerPolicy

>   name: tracing-config

Reviewer 2: This means we won't be able to support per-route tracing? https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-tracing If that's the case we should mention it in the non-goals section.

Author: Yes, added to the section. Thanks for pointing this out!

spec:
  tracing:
    openTelemetryConfig:
      grpcService:
        serviceName: tracing-service
        backendRef:
          name: gateway-proxy-tracing-server
          namespace: default
          port: 8083
        resourceDetectors:
          - EnvironmentResourceDetector
  accessLog:

>     - openTelemetryConfig:

Author: > Is OpenTelemtryConfig a new AccessLog value? 

Yes, since it does have additional fields. I'll try to extract the common fields (which right now is the log name and backendref) into a separate in the struct that can be reused in both configs (as a nameless field so it is not a breaking change)

        grpcService:
          logName: test-accesslog-service
          backendRef:
            name: gateway-proxy-access-logger
            namespace: default
            port: 8083
          body: '[%START_TIME%] "%PROTOCOL%" %RESPONSE_CODE%\n'
          attributes:
            values:
              - key: "user-agent"
                value:
                  string_value: "%REQ(USER-AGENT)%"
              - key: "duration"
                value:
                  double_value: "%DURATION%"
```

The `http_listener_policy_types.go` will be modified to add the following fields :
```go

> type HTTPListenerPolicySpec struct {

Reviewer 2: What's responsible for configuring https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#extensions-filters-network-http-connection-manager-v3-httpconnectionmanager-tracing?

The [OTel access log sink](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/access_loggers/open_telemetry/v3/logs_service.proto) provides greater flexibility to add metadata to access logs such as [attributes](https://opentelemetry.io/docs/specs/semconv/general/attributes/), [formatters](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/formatter/formatter), and providing a standard way to propagate and record logs.

## Goals

- Users should be able to configure OTel tracing via kgateway
- Users should be able to configure OTel access logging via kgateway
- Support tracing at a Gateway level
- Provide documentation and example configurations for integrating kgateway with popular observability platforms (e.g., Jaeger, Datadog)
- Maintain [existing access logging support](https://kgateway.dev/docs/security/access-logging/)

## Non-Goals

- Deploying an OTel Collector as part of the kgateway Helm chart
- Providing opinionated configurations with specific observability vendors or backends out of the box (integrations with kgateway should be vendor agnostic)

> - Support per-route tracing

Reviewer 2: Thanks for adding this to the non-goals section! Can we add a link to https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-tracing for future reference. At some point I'm guessing we'll need to revisit this and implement per-route 😅

Author: Updated

    SkipEnvoyHeaders *bool `json:"skipEnvoyHeaders,omitempty"`

    // The timeout for the gRPC request. This is the timeout for a specific request
    Timeout *metav1.Duration `json:"timeout,omitempty"`

    // Additional metadata to include in streams initiated to the GrpcService.
    // This can be used for scenarios in which additional ad hoc authorization headers (e.g. x-foo-bar: baz-key) are to be injected
    InitialMetadata []*HeaderValue `json:"initialMetadata,omitempty"`
}

// +kubebuilder:validation:MaxProperties=1
// +kubebuilder:validation:MinProperties=1
type ResourceDetector struct {
    EnvironmentResourceDetector EnvironmentResourceDetector `json:"environmentResourceDetector,omitempty"`
    // New detectors can be added as needed
}

// +kubebuilder:validation:MaxProperties=1
// +kubebuilder:validation:MinProperties=1
type Sampler struct {
    // This is just an example. New samplers can be added
    AlwaysOn AlwaysOn `json:"alwaysOn,omitempty"`
}

type EnvironmentResourceDetector struct {}
```
The existing AccessLog struct will be modified to add support for OTel access logs
```go
type AccessLog struct {
    // Send access logs to an OTel collector

>     OpenTelemetryConfig OpenTelemetryAccessLogConfig `json:"openTelemetryConfig,omitempty"`

Reviewer 2: [nit] Maybe we can leave off config here (just `openTelemetry`) to fit the structure of the other fields that don't use config:

```
// AccessLog represents the top-level access log configuration.
type AccessLog struct {
	// Output access logs to local file
	FileSink *FileSink `json:"fileSink,omitempty"`

	// Send access logs to gRPC service
	GrpcService *GrpcService `json:"grpcService,omitempty"`

	// Filter access logs configuration
	Filter *AccessLogFilter `json:"filter,omitempty"`

    // Send access logs to an OTel collector
    OpenTelemetryConfig *OpenTelemetryAccessLog `json:"openTelemetry,omitempty"`
}
```

Also will AccessLogFilter work with OpenTelemetryConfig?

Author: Good idea. updated it
Also AccessLogFilter should work but I can test it and update the docs accordingly

}

// OpenTelemetryAccessLogConfig represents the OTel configuration for access logs.
type OpenTelemetryAccessLogConfig struct {
    // Send access logs to gRPC service
    // +kubebuilder:validation:Required
    GrpcService *OTelGrpcService `json:"grpcService,omitempty"`

    // name of log stream
    // +kubebuilder:validation:Required
    LogName string `json:"logName"`

    // OpenTelemetry LogResource fields, following Envoy access logging formatting.
    // +kubebuilder:validation:Optional
    Body string `json:"body,omitempty"`

    // If specified, Envoy will not generate built-in resource labels like log_name, zone_name, cluster_name, node_name.
    // +kubebuilder:validation:Optional

>     DisableBuiltinLabels bool `json:"disableBuiltinLabels,omitempty"`

Reviewer 2: All omitempty fields should be pointers. We may also want to change some of these bools to enum types.

    // Additional attributes that describe the specific event occurrence.
    // +kubebuilder:validation:Optional
    Attributes KeyValueList `json:"attributes,omitempty"`
}
```

### Plugin

The existing `httplistenerpolicy` plugin performs the access log translation and can be updated to support OTel access log and tracing.

### Controllers

No new controllers will be required for this feature. The existing plugin will be updated to support this.

### Deployer

The CRDs will be updated with the new fields along with the required validation

### Test Plan

> The testing plan will include unit and kubernetes end to end tests to verify the feature

Author: @krisztianfekete mentioned that he would make it public. Until then I'll keep the manifest in the repo for tests

## Alternatives

> - The tracing config can be defined in the `GatewayExtension`, however this could lead to invalid configs where a user might inadvertently apply it where it is not supported, eg: an HTTPRoute.

Reviewer 2: But eventually we may want route support for tracing and access logs? So maybe it does make sense to have it configured on the GatewayExtension:
```
# gateway extension that will be applied  routes
apiVersion: gateway.kgateway.dev/v1alpha1
kind: GatewayExtension
metadata:
  name: basic-extauth
  namespace: infra
spec:
  type: Tracing
  tracing:
    grpcService:
        serviceName: tracing-service
        backendRef:
          name: gateway-proxy-tracing-server
          namespace: default
          port: 8083
```

@lgadban @yuval-k you have background on when to use GatewayExtension. Would it make sense here, or is it better to inline the config in the HTTPListenerPolicy.

Reviewer 3: We haven't been able to put together concrete guidance but the rule(s) of thumb to this point have been something like:
* is this required to configure a standalone envoy filter?
* does it define a standalone service or provider? (and provide associated config with that provider?)

The tracing config only is relevant to the HCM right? And same with ALS? If so, I think having it inline seems reasonable for now.
If there really becomes an option for more discrete config of access logging and tracing we can explore the GwExt API.

Reviewer 4: I"ll give the  envoy point of view. We don't have a good way to merge config that goes on the filter's listener config. so provider can be used to decouple route level and listener level configs.
I think this config never gets to a route, right? so I agree with @lgadban that inline should be ok

## Open Questions

> - Which detectors should be supported ?

Reviewer 2: Envoy detector? Let's link to https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/tracers/opentelemetry/resource_detectors/v3/environment_resource_detector.proto 

I think since it's an experimental extension, it might make sense to leave it out of scope for now?

Author: I think we can add support for the [Environment Resource Detector](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/tracers/opentelemetry/resource_detectors/v3/environment_resource_detector.proto#extension-envoy-tracers-opentelemetry-resource-detectors-environment) mentioned in https://github.com/kgateway-dev/kgateway/issues/9267 which is working on the dev branch

> - Should HTTP services be supported for tracing (https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/http_service.proto#envoy-v3-api-msg-config-core-v3-httpservice) ?

Reviewer 2: Is this question for supporting configuring insecure/secure? Or what is it referring to?

Author: It is referring to configuring tracing over (to an upstream) HTTP as well as gRPC service
